novawallet/Common/URLHandling/Parsing/OpenGovernanceUrlParsingService.swift

```diff
@@ -0,0 +1,74 @@
+import Foundation
+
+final class OpenGovernanceUrlParsingService: OpenScreenUrlParsingServiceProtocol {
+    private let chainRegistryClosure: ChainRegistryLazyClosure
+
+    enum Key {
+        static let chainid = "chainid"
+        static let referendumIndex = "id"
+        static let governanceType = "type"
+    }
+
+    init(chainRegistryClosure: @escaping ChainRegistryLazyClosure) {
+        self.chainRegistryClosure = chainRegistryClosure
+    }
+
+    func parse(url: URL) -> Result<UrlHandlingScreen, DeeplinkParseError> {
+        guard let urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false),
+              let query = urlComponents.queryItems else {
+            return .failure(.openGovScreen(.emptyQueryParameters))
+        }
+
+        let queryItems = query.reduce(into: [String: String]()) {
+            $0[$1.name.lowercased()] = $1.value ?? ""
+        }
+
+        guard let chainId = queryItems[Key.chainid],
+              !chainId.isEmpty else {
+            return .failure(.openGovScreen(.invalidChainId))
+        }
+
+        guard let referendumIndexString = queryItems[Key.referendumIndex],
+              let referendumIndex = UInt(referendumIndexString) else {
+            return .failure(.openGovScreen(.invalidReferendumId))
+        }
+
+        guard let chainModel = chainRegistryClosure().getChain(for: chainId) else {
```

Comment:
It is better to use subscription and handle matches in it. This will allow us to handle case when chains are out of sync. In this case parse function also must be async.